### ISSUE TITLE
"Don't expand collapsed tree and skip collapsed descendants" - not working properly

### ISSUE THREAD

**Issue by pintassilgo** [piroor/treestyletab#2442]
TST 3.2.6, Fx 71.0, Windows 10.

With this option checked (under Tree Behavior).
```
Tab A
Tab B
- Tab C (child of Tab B)
Tab D
```

First step: collapse Tab B tree.

First scenario:
I click Tab B, then press Ctrl+tab, Tab D is selected and the tree remains collapsed. OK.

Second scenario:
I click Tab A, then Ctrl+tab to Tab B. Then Ctrl+tab again. Now the tree is expanded and Tab C is selected. This is unwanted, the tree should remain collapsed, so Tab D should be selected.

Third scenario:
I click Tab D, then Ctrl+Shift+tab. The tree is expanded and Tab C is selected. This is unwanted, the tree should remain collapsed, so Tab B should be selected.

Fourth scenario:
I click Tab B, then press Ctrl+Shift+tab to Tab A, then again to Tab D, then again. Tab B is selected as it should be, but its tree is expanded and it shoudn't.

**Comment by piroor** [comment]
Sorry but please remind that the option does not work completely due to restrictions of WebExtensions API.

Basically Ctrl-Tab/Ctrl-Shift-Tab shortcuts are handled by Firefox itself not TST. TST just tries to redirect tab focus when the newly focused tab is a parent in a collapsed tree, or tries to keep the focused tree collapsed while the Ctrl key is pressed based con content scripts. This behavior is triggered with a keydown event fired in the content area or the sidebar, so, if you started tab switching from a special tab like `about:addons` or Mozilla Add-ons website (it disallows addons to run any content script) or any Firefox UI (the location bar, toolbar buttons, or others) is focused, no keydown event is tracked by TST and TST will fail to override Ctrl-Tab behavior of Firefox.

In other words, TST's Ctrl-Tab option works only with these both conditions:

* You started Ctrl-Tab/Ctrl-Shift-Tab on a regular webpage, not Mozilla Add-ons.
* You focused to the content area, not Firefox's UI.

Did you talk about any on these exceptional cases?

**Comment by pintassilgo** [comment]
I have AMO whitelisted by `extensions.webextensions.restrictedDomains = ""` and `privacy.resistFingerprinting.block_mozAddonManager = true`. But yes, there is no workaround for chrome pages like about:*.

But no, I did the tests in normal content pages, with focus on content or on TST sidebar. Yet the trees expanded when they shouldn't.

**Comment by pintassilgo** [comment]
Look:

![](https://i.imgur.com/zbUEYVb.gif)

Steps:
1. Clicked in the first tab (Inoreader, collapsed tree).
2. Pressed Ctrl+Tab.
3. The next root tab (WhatsApp) was selected, but after a short delay its tree was expanded and it shouldn't.
4. Pressed Ctrl+Shift+Tab.
5. The previous root tab was selected (OK), but after a short delay its tree was expanded too (unwanted).

None of the trees should have expanded in the process.

**Comment by piroor** [comment]
> The next root tab (WhatsApp) was selected, but after a short delay its tree was expanded and it shouldn't.

> The previous root tab was selected (OK), but after a short delay its tree was expanded too (unwanted).


Oh, you might told about an undocumented but intentional behavior of TST. TST is designed to expand the current tree while Ctrl-Tab/Ctrl-Shift-Tab focus rotation, when you keep the Tab key pressed for a while (800msec by default) on a collapsed tree. You can enlarge the delay with an internal option `autoExpandOnTabSwitchShortcutsDelay` and you can deactivate the behavior with `autoExpandOnTabSwitchShortcuts`. (You'll find these options under the section "Development" => "All Configs".)

**Comment by pintassilgo** [comment]
Thank you. I think `autoExpandOnTabSwitchShortcuts` should honor "Don't expand collapsed tree and skip collapsed descendants" or be a sub-setting of it.

A little off-topic: do you know how to access TST from chrome? I need to adjust my chrome-privileged mouse gesture (rightclick + wheel) to not expand collapsed tree on tab switching. Currently I use simply `gBrowser.tabContainer.advanceSelectedTab(-1, true);`. I can fix it myself, I just need a way to enter TST scope from chrome to check `tstTabs[previousTabIndex].states.includes('collapsed')`.

**Comment by piroor** [comment]
> Thank you. I think autoExpandOnTabSwitchShortcuts should honor "Don't expand collapsed tree and skip collapsed descendants" or be a sub-setting of it.

OK, I've exposed the option as an expert option with the commit 4e31ec183.

> do you know how to access TST from chrome?

Sorry, I don't know how to do that. Basically WebExtensions-based extensions are isolated and running in enclosed namespaces, so they are not accessible like as you wrote. You may find out some hints from [Firefox's automated tests around WebExtensions API](https://searchfox.org/mozilla-central/source/browser/components/extensions/test/browser) and a [common helper for them](https://searchfox.org/mozilla-central/source/toolkit/components/extensions/ExtensionXPCShellUtils.jsm#827) because some tests contain codes to run arbitrary scripts in a namespace for addons.
